File: pyAlgorithm/sort/merge_sort.py
import sys
import random
#input = sys.stdin.readline().rstrip() # 빠른 입력


#arr1 = random.sample(range(1,100),10)

arr1 = [24, 76, 70, 94, 20, 73, 38, 43, 98, 81]   
print(arr1)

#분할 파트
def merge_sort(arr):
    if len(arr) <= 1: # 종료 조건: arr 길이가 1일때 종료
        return arr

    mid = len(arr) // 2
    # 배열 슬라이싱: 깊은 복사로 배열을 복사
    # arr[a:b] -> a ~ (b-1) 인덱스까지 복사
    print(arr[:mid], arr[mid:])
    return merge(merge_sort(arr[:mid]), merge_sort(arr[mid:]))

#정복 파트 
def merge(left, right): 
    result = []
    i = j = 0
    print("-------")
    print("merge: ",left, right)
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # 더이상 한쪽 배열에 비교 대상이 없을 때,
    # 배열 안에 n개는 정렬이 되어있는 상태이므로 순서대로 넣어줌. 
    while i < len(left):
        result.append(left[i])
        i += 1

    while j < len(right):
        result.append(right[j])
        j += 1

    print("result: ",result)
    print("-------")

    return result

sorted_arr = merge_sort(arr1)

print(sorted_arr)
    

File: pyAlgorithm/sort/merge_sort.cpp
#include <iostream>
using namespace std;

int N,arr[1000001];
int *arr2;

void merge(int left, int right)
{
    printf("merge: %d %d\n",left, right);
	int mid = (left + right) / 2;

	int i = left; //arr 왼쪽배열 커서값
	int j = mid + 1; //arr 오른쪽 배열 커서값

	int k = left; //arr2 배열 커서 값
	while (i <= mid && j <= right)
	{
		if (arr[i] <= arr[j]) 
			arr2[k++] = arr[i++]; 
		else
			arr2[k++] = arr[j++];
	}

    //비교대상 없을 시,
	int tmp = i > mid ? j : i;
	
	while(k<=right) arr2[k++] = arr[tmp++];

    //arr2에 있는 값 arr에 복사
	for (int i=left; i<=right;i++){
        arr[i] = arr2[i];
        printf("%d ",arr[i]);
    } 
    printf("\n");
}

void merge_sort(int left,int right)
{
	int mid;
	if (left < right)
	{
		mid = (left + right) / 2; 
		merge_sort(left, mid);
		merge_sort(mid + 1, right);

        printf("%d %d\n",left, right);
		merge(left, right);
	}
}

int main()
{
	scanf("%d",&N);
	arr2 = new int[N];
	for (int i=0;i<N;i++) scanf("%d",&arr[i]);

	merge_sort(0, N - 1);

	for (int i=0;i<N;i++) printf("%d ",arr[i]) ;
}